src/Lecture4.hs

```diff
@@ -177,196 +73,76 @@ data Stats = Stats
-   you can set the corresponding field to 'Nothing'.
--}
+parseRow :: String -> Maybe Row
+parseRow "" = Nothing
+parseRow rStr = if ((null $ splitRow (== ',') rStr) || ((listLen $ splitRow (== ',') rStr) /= 3)) then Nothing
+	            else
+                  do
+                     let (p:t:c:_) = splitRow (== ',') rStr
+                     t <- readMaybe t:: Maybe TradeType
+                     c <- costCheck(readMaybe c:: Maybe Int)
+                     return (Row { rowProduct = p, rowTradeType = t, rowCost = c})	
+            
+splitRow :: (Char -> Bool) -> String -> [String]
+splitRow p s =  case dropWhile p s of
+                      "" -> []
+                      s' -> w : splitRow p s''
+                            where (w, s'') = break p s'
+
+costCheck :: Maybe Int -> Maybe Int
+costCheck a = if ((fmap (< 0) a)== Just True) then Nothing else a
+
+listLen :: [String] -> Int
+listLen [] = 0
+listLen ((x:xs):ys) = length $ fmap length ((x:xs):ys)
 
 rowToStats :: Row -> Stats
-rowToStats = error "TODO"
-
-{-
-Now, after we learned to convert a single row, we can convert a list of rows!
-
-However, we have a minor problem. Our 'Stats' data type doesn't have a
-'Monoid' instance and it couldn't have it! One reason for this is that
-there's no sensible "empty" value for the longest product name. So we
-simply don't implement the 'Monoid' instance for 'Stats'.
-
-But the list of rows might be empty and we don't know what to return
-on empty list!
-
-The solution of this problem is to propagate handling of this
-situation upstream. In our type signature we will require to accept a
-non-empty list of rows.
-
-We can use the 'NonEmpty' data type from the 'Data.List.NonEmpty'
-module for this purpose. 'NonEmpty' is like 'List1' from Lecture 3
-exercises (remember that type?) but with the different constructor.
-
-Have a look at the 'sconcat' function from the 'Semigroup' typeclass to
-implement the next task.
--}
+rowToStats (r@(Row p tt c)) 
+    | tt == Buy = Stats { statsTotalPositions = Sum 1, statsTotalSum = Sum (-c), statsAbsoluteMax = Max (c), statsAbsoluteMin = Min (c), statsSellMax = Nothing, statsSellMin = Nothing, statsBuyMax = Just (Max (c)), statsBuyMin = Just (Min (c)), statsLongest = MaxLen (p)}
+	| tt == Sell = Stats { statsTotalPositions = Sum 1, statsTotalSum = Sum (c), statsAbsoluteMax = Max (c), statsAbsoluteMin = Min (c), statsSellMax = Just (Max (c)), statsSellMin = Just (Min (c)), statsBuyMax = Nothing, statsBuyMin = Nothing, statsLongest = MaxLen (p)}
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
-
-{-
-After we've calculated stats for all rows, we can then pretty-print
-our final result.
-
-If there's no some value (for example, there were not "Buy" products),
-you can return string "no value"
--}
+combineRows (r:|[]) = rowToStats r
+combineRows (r:|rs) = let (s:|st) = head (fmap rowToStats (r:rs)):|tail (fmap rowToStats (r:rs)) in sconcat (s:|st)
```

Comment:
Your idea is correct but you can simplify the implementation of this function by using the fact that `NonEmpty` has `Functor` instance:

```suggestion
combineRows = sconcat . fmap rowToStats
```